### ISSUE TITLE
DX: using `client` as the route name causes an error

### ISSUE THREAD

**Issue by navid-asgharzadeh** [trpc/trpc#3710]
### Provide environment information

```
System:
    OS: macOS 13.1
    CPU: (8) x64 Apple M1
    Memory: 18.31 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
    Yarn: 1.22.10 - /opt/homebrew/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
    Watchman: 2022.10.03.00 - /opt/homebrew/bin/watchman
  Browsers:
    Brave Browser: 109.1.47.171
    Chrome: 109.0.5414.119
    Safari: 16.2
  npmPackages:
    @tanstack/react-query: ^4.20.4 => 4.24.4 
    @trpc/client: ^10.7.0 => 10.9.1 
    @trpc/next: ^10.7.0 => 10.9.1 
    @trpc/react-query: ^10.7.0 => 10.9.1 
    @trpc/server: ^10.7.0 => 10.9.1 
    next: ^13.1.1 => 13.1.6 
    react: ^18.2.0 => 18.2.0 
    typescript: ^4.9.4 => 4.9.5 
```

### Describe the bug

I've created a route and named it `client` but I get this error when trying to use `trpc.useContext()`

```.ts
client collides with a built-in method, you should rename this router or procedure on your backend
```
I could fix the issue by renaming it to `clientRoute` and the problem was solved

### Link to reproduction

https://stackblitz.com/github/trpc/examples-next-minimal-starter

### To reproduce

Create a route and name it `client`, This should trigger an error as shown in the video below:

https://user-images.githubusercontent.com/56588770/216343859-4bd76863-845e-45bd-a664-0911b4ed0054.mov

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

<sub>[TRP-54](https://linear.app/trpc/issue/TRP-54/dx-using-client-as-the-route-name-causes-an-error)</sub>

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/3710">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/3710/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/3710/pledge.svg">
</picture>
</a>

**Comment by KATT** [comment]
I think this happened after upgrading as a result of #3693.

I think it's worth investigating a bit more if there are non-breaking workarounds we can do to prevent this as I think it's really bad DX

**Comment by KATT** [comment]
- For the context proxy, I think we need to separate the `useContext()` that contains the extra properties themselves 
  > Example: `trpc.useContext().post.byId.invalidate()` and `trpc.useSomethingElse().client`
- For `createTRPCNext()` I think we're good for now, it seems unlikely you'd name something `useContext`/`useQueries`/`withtRPC`
- For `createTRPCReact()` I think we're fine too.

This does look messy though and smells like the design could be improved:
<img width="623" alt="CleanShot 2023-02-02 at 09 22 12@2x" src="https://user-images.githubusercontent.com/459267/216350328-19fc437d-f314-43c3-881e-4a1db791d00d.png">

**Comment by websideofamir** [comment]
HI @KATT 
i'm trying to build my nextjs app and getting the fooowing error. is this related?
`
Type error: Property 'serverGroup' does not exist on type '"useContext collides with a built-in method, you should rename this router or procedure on your backend" | "withTRPC collides with a built-in method, you should rename this router or procedure on your backend" | "useQueries collides with a built-in method, you should rename this router or procedure on your backend"'.
  Property 'serverGroup' does not exist on type '"useContext collides with a built-in method, you should rename this router or procedure on your backend"'.
`

**Comment by asterikx** [comment]
I'm also getting this error randomly on my build server:

```
app:build: Failed to compile.
--
app:build:
app:build: ./src/pages/_app.tsx:87:21
app:build: Type error: Property 'withTRPC' does not exist on type '"The property 'useContext' in your router collides with a built-in method, rename this router or procedure on your backend." \| "The property 'withTRPC' in your router collides with a built-in method, rename this router or procedure on your backend." \| "The property 'useQueries' in your router collides with a built-i...'.
app:build:   Property 'withTRPC' does not exist on type '"The property 'useContext' in your router collides with a built-in method, rename this router or procedure on your backend."'.
app:build:
app:build:   86 \|
app:build: > 87 \| export default trpc.withTRPC(App);
app:build:      \|                     ^
app:build:   88 \|
app:build:  ELIFECYCLE  Command failed with exit code 1.
app:build: ERROR: command finished with error: command (/vercel/path0/packages/app) pnpm run build exited (1)
command (/vercel/path0/packages/app) pnpm run build exited (1)
 
Tasks:    1 successful, 2 total
Cached:    1 cached, 2 total
Time:    19.473s
Summary:    /vercel/path0/.turbo/runs/2OL0f9u0IvChFcvtJWdcgOBVjIw.json
 
ERROR  run failed: command  exited (1)
ELIFECYCLE  Command failed with exit code 1.
Error: Command "pnpm build:app" exited with 1
Deployment completed
BUILD_UTILS_SPAWN_1: Command "pnpm build:app" exited with 1
```

**Comment by devidw** [comment]
Hitting the same error when trying to construct via `createTRPCNext`:

![Screenshot 2023-06-15 at 17 21 29](https://github.com/trpc/trpc/assets/68775926/fc6552b6-f736-470c-957c-e2c41636a695)

Searching for those bad names in the whole codebase doesn't show of any usage of those in our backend routers

On version `10.30.0`

Tried to downgrade to before `10.9.1` as menitoned [here](https://github.com/trpc/trpc/issues/3710#issuecomment-1413795302) as possible source

Going for `10.9.0` just gives the same error message, just slightly different:

```
"useContext collides with a built-in method, you should rename this router or procedure on your backend"
```

---

Not sure if that's related, but we are trying to set up a connection between a next.js app and a standalone, separately hosted trpc server.  We don't get this error when constructing `createTRPCProxyClient` within a SvelteKit and `createTRPCReact` within React Native projects with the same `AppRouter` type.

**Comment by kasokuta** [comment]
I am having this issue, any news on solving this problem?

**Comment by lucasrabiec** [comment]
Bump. Same in v10.35.0.

**Comment by naviocean** [comment]
I got the same issue. :(

**Comment by touyu** [comment]
I have the same problem.
Do you have any plans to resolve it?

**Comment by FleetAdmiralJakob** [comment]
Have the same problem. What can I do?

**Comment by KATT** [comment]
Rename it to something else.

**Comment by FleetAdmiralJakob** [comment]
> Rename it to something else.

Hey, what exactly should I rename?
I fixed the error with updating tRPC

**Comment by alemanoelsilva** [comment]
Hey @FleetAdmiralJakob 

Which version are you using?

I've got the same issue, I tried to update the tRPC to the most recent version (10.38.0), but for me, the error persists.

**Comment by vogtb** [comment]
I'm experiencing this on every version except `10.7.0`.

**Comment by sachinraja** [comment]
Does anyone have a reproduction where this error should not be showing up (you do not have actually have a procedure named something that collides with a built-in method)?

**Comment by HiepPP** [comment]
Encountered the same issue. Downgrading to 10.30.0 resolved it temporarily.

**Comment by shakhzodkudratov** [comment]
explicitly setting type helped me to resolve this annoying situation

instead of
```ts
export const trpc = createTRPCReact<AppRouter>()
```

i wrote
```ts
export const trpc: CreateTRPCReact<AppRouter, unknown, null> = createTRPCReact<AppRouter>()
```

**Comment by shakhzodkudratov** [comment]
as it turned out, my solution above was not solving the situation at all. After killing a lot of time doing research, I finally understand that all the problem was because of references in tsconfig.json file

https://github.com/kuubson/react-vite-trpc this repository helped me

tldr;
add the following inside the root object of tsconfig.json in the project where you want to import AppRouter:
```json
{
    "compilerOptions": {...},
    "references": [
          {
              "path": "../server" // path to the project's tsconfig.json which has AppRouter type we want to import
          }
    ]
}
```

**Comment by braden-w** [comment]
The root cause in my case was incorrect Typescript path aliases in the server project. Switching to relative paths worked.

I noticed that the `AppRouter` type was being inferred as `any` when checking type definitions (even when I inspected it where it was defined in my server project).

The fix for me was to temporarily remove Typescript aliases on the server side, use relative paths for everything, and everything worked once again!

**Comment by suyu0925** [comment]
I also encountered this issue, and later discovered that the reason was a silly mistake I made. I didn't run `npm i` in the `server directory`.

**Comment by qwertzguy** [comment]
I had the same issue and for me the fix was to delete the node_modules **and** the package-lock.json from the root folder of my repo and the two folders for my app and server. Then ran npm install only in the root. That made it so I only have a node_modules and package-lock.json in the root folder and no longer in the two subfolders and made the error go away.
